tests/lifecycle_wait_test.go

```diff
@@ -14,366 +14,546 @@ import (
 	// Test description
 	//
 	// 1. goto /home and wait for the networkidle lifecycle event.
-	// 2. use WaitForLoadState with networkidle to ensure that
-	//    networkidle lifecycle event has already fired.
+	// 2. use WaitForNavigation with networkidle.
 	//
-	// Success criteria: We wait for all network requests and async
-	//                   scripts to complete. We also want to ensure
-	//                   that the networkidle event is stored internally,
-	//                   and we don't block on WaitForLoadState.
-
-	t.Parallel()
+	// Success criteria: Time out reached after navigation completed and
+	//                   wait for lifecycle event set, to signify that
+	//                   WaitForNavigation must be set before we navigate
+	//                   to a new page.
 
 	tb := newTestBrowser(t, withFileServer())
 	p := tb.NewPage(nil)
-	tb.withHandler("/home", func(w http.ResponseWriter, r *http.Request) {
-		http.Redirect(w, r, tb.staticURL("wait_for_nav_lifecycle.html"), http.StatusMovedPermanently)
-	})
-
-	var counter int64
-	var counterMu sync.Mutex
-	tb.withHandler("/ping", func(w http.ResponseWriter, _ *http.Request) {
-		counterMu.Lock()
-		defer counterMu.Unlock()
 
-		counter++
-		fmt.Fprintf(w, "pong %d", counter)
-	})
-
-	tb.withHandler("/ping.js", func(w http.ResponseWriter, _ *http.Request) {
-		fmt.Fprintf(w, `
-				var pingJSTextOutput = document.getElementById("pingJSText");
-				pingJSTextOutput.innerText = "ping.js loaded from server";
-			`)
-	})
+	withHomeHandler(t, tb, "prolonged_network_idle_10.html")
+	withPingHandler(t, tb, 0, nil)
 
 	waitUntil := common.LifecycleEventNetworkIdle
-	assertHome(t, tb, p, waitUntil, func() {
-		result := p.TextContent("#pingRequestText", nil)
-		assert.EqualValues(t, "Waiting... pong 10 - for loop complete", result)
+	var resolved, rejected bool
+	err := tb.await(func() error {
+		opts := tb.toGojaValue(common.FrameGotoOptions{
+			WaitUntil: waitUntil,
+			Timeout:   30 * time.Second,
+		})
+		prm := tb.promise(p.Goto(tb.URL("/home"), opts)).then(
+			func() testPromise {
+				result := p.TextContent("#pingRequestText", nil)
+				assert.EqualValues(t, "Waiting... pong 10 - for loop complete", result)
 
-		result = p.TextContent("#pingJSText", nil)
-		assert.EqualValues(t, "ping.js loaded from server", result)
+				waitForNav := p.WaitForNavigation(tb.toGojaValue(&common.FrameWaitForNavigationOptions{
+					Timeout:   1000,
+					WaitUntil: waitUntil,
+				}))
+
+				return tb.promise(waitForNav)
+			},
+		)
+		prm.then(
+			func() {
+				resolved = true
+			},
+			func() {
+				rejected = true
+			},
+		)
 
-		// This shouldn't block and return after calling hasLifecycleEventFired.
-		p.WaitForLoadState(waitUntil.String(), nil)
+		return nil
 	})
+	require.NoError(t, err)
+
+	assert.False(t, resolved)
+	assert.True(t, rejected)
 }
 
-func TestLifecycleWaitForLoadStateDOMContentLoadedThenNetworkIdle(t *testing.T) {
+func TestLifecycleWaitForLoadState(t *testing.T) {
```

Comment:
Can we put these tests into `TestLifecycleWaitForNavigation`? Optional: To do that, maybe we can create a `home` struct and have some methods. I believe fewer things to read and less code are mostly better regarding comprehensibility and maintenance. I think `assertHome` deserves a better home.

Reply:
I can take a look at making `assertHome` more generic.

All these separate tests functions test different methods (`goto`, `waitForNavigation`, `waitForLoadState`, `reload`), and although each of them use `waitUntil` in a similar way, they are still different methods. If we accidentally change one of these, then only the related test function will break, which should be easier to understand.

`goto` is tested in a few different ways, one is with `subFrames`, another is without `subFrames` and finally we test `networkIdle` in different scenarios where the website is different enough whereby the lifecycle events are emitted by the browser (subtly) differently -- i.e. they are edge cases. I think they should also live in their own functions.